### PR TITLE
Legends Awakened Bugfixes

### PR DESCRIPTION
**Pull request by luongthomas**
Addressing issues regarding the Gen4 Legends Awakened set bugs found on trello

### CODE REVIEW

**Comment by luongthomas** [comment]
[KYOGRE_32](https://trello.com/c/DVNVyDZl/1137-kyogre32)

**Comment by luongthomas** [comment]
[LOPUNNY_33](https://trello.com/c/wNvYrkbP/1138-lopunny33)

**Comment by luongthomas** [comment]
[REGICE_36](https://trello.com/c/u8FFtzlh/1141-regice36)

**Comment by luongthomas** [comment]
[REGISTEEL_39](https://trello.com/c/pQq7ZTa1/1144-registeel39)

**Comment by luongthomas** [comment]
[SHEDINJA_40](https://trello.com/c/gg8oY9Rp/1145-shedinja40)

**Comment by luongthomas** [comment]
[VICTREEBEL_44](https://trello.com/c/d9lZJ2L1/1149-victreebel44)

**Comment by luongthomas** [comment]
[VILEPLUME_45](https://trello.com/c/BhEyJL31/1150-vileplume45)

**Comment by luongthomas** [comment]
[ANORITH_46](https://trello.com/c/YAYqKSja/1151-anorith46)

**Comment by luongthomas** [comment]
[EXEGGUTOR_54](https://trello.com/c/58GPwWjY/1159-exeggutor54)

**Comment by luongthomas** [comment]
[HOUNDOOM_57](https://trello.com/c/pdzeD6uL/1162-houndoom57)

**Comment by luongthomas** [comment]
[LEDIAN_60](https://trello.com/c/9svAtGV2/1165-ledian60)

**Comment by luongthomas** [comment]
[LUCARIO_61](https://trello.com/c/TdTE8fgP/1166-lucario61)

**Comment by luongthomas** [comment]
[NINJASK_67](https://trello.com/c/9hwCPqcw/1172-ninjask67)

**Comment by luongthomas** [comment]
[TAUROS_74](https://trello.com/c/KDCQsXNO/1179-tauros74)

**Comment by luongthomas** [comment]
[TENTACRUEL_75](https://trello.com/c/kIXtUHuH/1180-tentacruel75)

**Comment by luongthomas** [comment]
[UNOWN_J_76](https://trello.com/c/TYhys635/1181-unownj76)

**Comment by luongthomas** [comment]
[UNOWN_R_77](https://trello.com/c/X8osSlWW/1182-unownr77)

**Comment by luongthomas** [comment]
[UNOWN_U_78](https://trello.com/c/AqwTVvh7/1183-unownu78)

**Comment by luongthomas** [comment]
[BELDUM_83](https://trello.com/c/j8B2XXap/1188-beldum83)

**Comment by luongthomas** [comment]
[BELLSPROUT_85](https://trello.com/c/n6AayR6O/1190-bellsprout85)

**Comment by luongthomas** [comment]
[SNEASEL_120](https://trello.com/c/zWOExvAe/1225-sneasel120)

**Comment by luongthomas** [comment]
[CHINCHOU_87](https://trello.com/c/dAGzyuMC/1192-chinchou87)

**Comment by luongthomas** [comment]
[POKE_RADAR_133](https://trello.com/c/c32RiKAU/1238-pokeradar133)

**Comment by luongthomas** [comment]
[MEWTWO_LV_X_144](https://trello.com/c/xTBaPeje/1249-mewtwolvx144)

**Comment by luongthomas** [comment]
[MESPRIT_LV_X_143](https://trello.com/c/IIDcmX45/1248-mespritlvx143)

**Comment by luongthomas** [comment]
[UNOWN_Y_81](https://trello.com/c/jXBkahU1/1186-unowny81)

**Comment by luongthomas** [comment]
[POLIWRATH_35](https://trello.com/c/N4ZkvfQT/1140-poliwrath35)

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 1]
```diff
@@ -2349,7 +2357,7 @@ public enum LegendsAwakened implements LogicCardInfo {
             energyCost G
             attackRequirement {}
             onAttack {
-              while(opp.hand.size()>4) {
+              while(opp.hand.size() > 5) {
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:2360]: Clearer to include the number used in the card text in the code.

```suggestion
              while(opp.hand.size() >= 6) {
```

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 2]
```diff
@@ -4119,14 +4136,14 @@ public enum LegendsAwakened implements LogicCardInfo {
             text "Prevent all effects of attacks, including damage, done to Mewtwo by your opponent's Pokémon that isn't an Evolved Pokémon."
             delayedA {
               before null, self, Source.ATTACK, {
-                if (self.owner.opposite.pbg.active.evolution && bg.currentTurn==self.owner.opposite && ef.effectType != DAMAGE){
```
Comment (starg09) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:4122]: Not necessarily specific to this card: How does this work with Lv.X basics? Also, basics with a card on top (not relevant to worlds formats, but there's Ditto BCR which could cause issues in unlimited/HoF)

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 3]
```diff
@@ -2412,8 +2420,8 @@ public enum LegendsAwakened implements LogicCardInfo {
             onAttack {
               damage 30
 
-              if (self.cards.filterByEnergyType(W)) {
-                def toDiscard = self.cards.filterByEnergyType(W).select(min: 0, max: 2, "For each [W] Energy discarded, do +10")
+              if (my.hand.filterByBasicEnergyType(W)) {
+                def toDiscard = my.hand.filterByBasicEnergyType(W).select(min: 0, max: 2, "For each [W] Energy discarded, do +10")
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:2424]: It does not state "Basic Energy cards." Theoretically if a Special Energy card is released that provides [W] by default, this attack should be able to discard it.

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 4]
```diff
@@ -2549,7 +2557,7 @@ public enum LegendsAwakened implements LogicCardInfo {
             attackRequirement {}
             onAttack {
               if (opp.discard.filterByType(ENERGY)) {
-                def energies = opp.discard.filterByType(ENERGY).select("Select up to 2 Energy Cards to attach")
+                def energies = opp.discard.filterByType(ENERGY).select(min:0, max: 2, "Select up to 2 Energy Cards to attach")
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:2560]: Up to means 1 or more. Not 0. May is the keyword used for 0 or more.

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 5]
```diff
@@ -4119,14 +4136,14 @@ public enum LegendsAwakened implements LogicCardInfo {
...
                   prevent()
                 }
               }
               before APPLY_ATTACK_DAMAGES, {
                 bg.dm().each {
-                  if (it.to == self && it.notNoEffect && it.from.evolution) {
+                  if (it.to == self && it.notNoEffect && !it.from.evolution) {
                     it.dmg = hp(0)
                     bc "Psybarrier prevents damage"
                   }
```
Comment (starg09) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:4149]: Same issue as the other comment on the card, but also adding an "opponent's pokemon" check:

```suggestion
                if (ef.attacker.owner != self.owner && ef.attacker.evolution) {
                  bg.dm().each {
                    if (it.to == self && it.dmg.value && it.notNoEffect) {
                      it.dmg = hp(0)
                      bc "Psybarrier prevents damage"
                    }
                  }
```

**Review by starg09 — COMMENTED** [review]
Looks good, left some comments :)

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 6]
```diff
@@ -2788,12 +2802,12 @@ public enum LegendsAwakened implements LogicCardInfo {
             text "Once during your turn (before your attack), when you attach a [M] Energy card from your hand to Beldum (excluding effects of attacks or Poké-Powers), you may search your deck for Beldum and put it onto your Bench. Shuffle your deck afterward. This power can't be used if Beldum is affected by a Special Condition."
             delayedA {
               before ATTACH_ENERGY, {
-                if (ef.reason == PLAY_FROM_HAND && ef.card.asEnergyCard().containsType(W) && bg.currentTurn == self.owner && ef.resolvedTarget == self && self.noSPC()) {
+                if (ef.reason == PLAY_FROM_HAND && ef.card.asEnergyCard().containsType(M) && bg.currentTurn == self.owner && ef.resolvedTarget == self && self.noSPC() && confirm("Use Metal Chain?")) {
                   checkLastTurn()
                   assert my.bench.notFull : "Bench full"
                   assert deck.notEmpty : "Deck is empty"
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:2808]: These should be in the if statement. No reason to ask the player "want to use this ability?" "too bad you can't"

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 7]
```diff
@@ -4102,7 +4115,11 @@ public enum LegendsAwakened implements LogicCardInfo {
             text "200 damage. If you don't have Uxie LV.X and Azelf LV.X in play, this attack does nothing. Discard all Energy attached to Mesprit."
             energyCost P, P
             attackRequirement {
-              my.all.findAll { it.name == ("Uxie Lv.X") } && my.all.findAll{ it.name == ("Azelf Lv.X") }
+              assert my.all.findAll {
+                it.name == ("Uxie") && it.pokemonLevelUp
+              } && my.all.findAll {
+                it.name == ("Azelf") && it.pokemonLevelUp
+              } : "Uxie Lv.X and Azelf Lv.X are not in play"
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:4122]: `findAll` isn't necessary, `find` or `any` should be enough.

**Review by flagrama — CHANGES_REQUESTED** [review]

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 8]
```diff
@@ -1624,7 +1625,7 @@ public enum LegendsAwakened implements LogicCardInfo {
                   before BETWEEN_TURNS, {
                     if (bg.currentTurn == self.owner.opposite) {
                       bc "Energy Dissolve activates"
-                      discardSelfEnergyAfterDamage C
+                      discardSelfEnergy C
```
[the remark this answers is not in the document]
Comment (luongthomas) [reply src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:1627]: I'm actually not sure how to solve this one and I will move the associated trello card to backlog

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 9]
```diff
@@ -1663,14 +1664,16 @@ public enum LegendsAwakened implements LogicCardInfo {
...
+                  before PLAY_TRAINER, {
+                    if (bg.currentTurn == self.owner.opposite && (ef.cardToPlay.cardTypes.is(TRAINER) || ef.cardToPlay.cardTypes.is(SUPPORTER) || ef.cardToPlay.cardTypes.is(STADIUM))) {
+                      wcu "Disturbing Pollen prevents playing Trainer Cards this turn"
+                      prevent()
+                    }
                   }
+                  unregisterAfter 2
                 }
-                unregisterAfter 2
               }
```
[the remark this answers is not in the document]
Comment (luongthomas) [reply src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:1676]: I've changed it to only look for TRAINERs

**Pull request merged**